File: (1) array/code_notes/traverse_from_right/py/daily_temperature.py
class Solution:
    # def dailyTemperatures(self, T: list[int]) -> list[int]:
    #     n, right_max = len(T), float('-inf') # n = length/size of T, right_max lowest possible float
    #     res = [0] * n  # creates a list of 0's with lenght n
    #     for i in range(n-1, -1, -1): # starting from len(T) - 1 (last index), stop at -1 (first index), skip by -1 (step from last element to first element)
    #         t = T[i] # t is the current element being pointed to
    #         if right_max <= t: # is right_max less than or equal to t?
    #             right_max = t # set right_max as t if true
    #             print(right_max)
    #         else:
    #             days = 1 
    #             while T[i+days] <= t: # while the next index after T[i] is less than or equal to t (T[i])
    #                 days += res[i+days] # increment and assign days to the value of the next index
    #             res[i] = days # assign days to result[i] (element at index i of res)
    #     return res
    def dailyTemperatures(self, T: list[int]) -> list[int]:
        # keep track of numbers and their index
        track: list[dict] = []
        n = len(T)
        res = [0] * n
        # print(next(iter(track[0].items()))[0])

        for i in range(n - 1, -1, -1):
            curr = T[i]
            while len(track) != 0 and next(iter(track[0].items()))[1] <= curr:
                track.pop(0)
                
            res[i] = 0 if len(track) == 0 else next(iter(track[0])) - i
            track.insert(0, {i: curr})
            # while len(track) != 0 and next(iter(track)) <= curr:
        return res

# i = i + step
print(Solution.dailyTemperatures(Solution, [73, 74, 75, 71, 69, 72, 76, 73]))

File: (1) array/code_notes/traverse_from_right/dart/daily_temperature.dart
void main() {
  print(dailyTemperatures([73, 74, 75, 71, 69, 72, 76, 73]));
}

List<int> dailyTemperatures(List<int> T) {
  // keep track of numbers seen when you move left
  var list = <Map<int, int>>[];
  int n = T.length;
  var res = List.filled(n, 0);

  for (int i = n - 1; i >= 0; i--) {
    int curr = T[i];
    while (list.isNotEmpty && list.first.values.first <= curr) {
      list.removeAt(0);
    }
    res[i] = list.isEmpty ? 0 : list.first.keys.first - i;
    list.insert(0, {i: curr});
  }

  return res;
}

// Map<int, int> m = {1 : 1};
// List<int> dailyTemperatures(List<int> T) {
//   var n = T.length;
//   var right_max = -1.minInt;
//   List<int> result = List.filled(n, 0);

//   for (int i = n - 1; i > 0; i--) {
//     var t = T[i];
//     if (right_max <= t) {
//       right_max = t;
//     } else {
//       int days = 1;
//       while (T[i + days] <= t) {
//         days += result[i + days];
//       }
//       result[i] = days;
//     }
//   }
//   return result;
// }

// extension on int {
//   int get minInt =>
//       (double.infinity is int) ? -double.infinity as int : (-1 << 63);
// }

// List<int> dailyTemperatures(List<int> temperatures) {
//   var p1 = temperatures.length - 1;
//   var p2 = temperatures.length - 1;
//   var result = [];

//   if (p1 == temperatures.length - 1) {
//     result.add(0);
//     p1 -= 1;
//   }

//   while (p1 >= 0) {
//     // while (p1 < p2) {
//     // if (temperatures[p1] < temperatures[p2]) {
//     //   print(true);
//     //   p2--;
//     // } else if (temperatures[p1] > temperatures[p2]) {
//     //   result.insert(0, p2 + 1);
//     //   p1--;
//     //   print(result);
//     // } else {
//     //   result.insert(0, 0);
//     //   print(result);
//     //   p1--;
//     // }
//     // while (p2 >= p1) {
//     //   if (temperatures[p1] < temperatures[p2]) {
//     //     p2--;
//     //   } else {
//     //     result.insert(0, p2 + 1);
//     //   }
//     // }
//     // p1--;
//     while (p2 > p1) {
//       if (temperatures[p1] > temperatures[p2]) {
//         result.insert(0, p2 - p1);
//         // print('${temperatures[p2] - temperatures[p1]}');
//         print('on if $p2 - $p1');
//       }
//       // else if (temperatures[p1] < temperatures[p2]) {
//       //   result.insert(0, 0);
//       // }
//       else {
//         result.insert(0, 0);
//       }
//       p2--;
//     }
//     // p2 = temperatures.length - 1;
//     p1--;
//   }
//   // p1--;
//   // }
//   print(result);
//   return [];
// }
